docs/guides/deno/byo-deep-research.mdx

```diff
@@ -0,0 +1,738 @@
+```
+
+> Code snippets presented will have the name of the file they are in included in the title.
+
+## The Architecture
+
+Our research system will be built as a Nitric API with several key components:
+
+1. **Search Engine Integration**
+2. **Configurable LLM Interface**
+3. **Topic-based Research Pipeline**
+4. **Content Processing**
+5. **REST API Endpoints**
+
+Let's break down each component:
+
+### 1. Search Engine Integration
+
+The search engine integration handles the "Web Search" step of the research pipeline:
+
+- Query Creation: Generate search queries from topics
+- Web Search: Find relevant content
+- Content Processing: Clean and convert content
+
+```typescript title:utils/search.ts
+import { search, SafeSearchType } from 'duck-duck-scrape'
+
+export default async (query: string) => {
+  const results = await search(query, { safeSearch: SafeSearchType.STRICT })
+
+  // Get the top 3 most relevant results
+  const topResults = results.results.slice(0, 3)
+
+  // Fetch HTML content for each result
+  const htmlContents = await Promise.all(
+    topResults.map(async (result) => {
+      try {
+        const response = await fetch(result.url)
+        const html = await response.text()
+        return {
+          url: result.url,
+          title: result.title,
+          html,
+        }
+      } catch (error) {
+        console.error(`Failed to fetch ${result.url}:`, error)
+        return null
+      }
+    }),
+  )
+
+  // Filter out failed fetches and return the results
+  return htmlContents.filter((content) => content !== null)
+}
+```
+
+### 2. Configurable LLM Integration
+
+The LLM integration handles the "Summarization" and "Reflection" steps:
+
+- Summarization: Condense findings
+- Reflection: Identify knowledge gaps
+- Iteration: Continue research as needed
+
+The system can be configured to use different LLM providers through environment variables:
+
+```typescript title:services/api.ts
+import { OpenAI } from 'openai'
+
+// Configuration from environment variables
+const LLM_CONFIG = {
+  baseURL: Deno.env.get('LLM_BASE_URL') || 'http://localhost:11434/v1',
+  apiKey: Deno.env.get('LLM_API_KEY') || 'ollama',
+  model: Deno.env.get('LLM_MODEL') || 'llama3.2:3b',
+}
+
+const OAI = new OpenAI({
+  baseURL: LLM_CONFIG.baseURL,
+  apiKey: LLM_CONFIG.apiKey,
+})
+
+// Example .env file:
+// LLM_BASE_URL=http://localhost:11434/v1   # For local testing with Ollama
+// LLM_BASE_URL=https://api.openai.com/v1   # For production with OpenAI
+// LLM_API_KEY=your-api-key                 # For local this can be any value
+// LLM_MODEL=gpt-4-turbo-preview            # For production (or any other model of your choosing)
+// LLM_MODEL=llama3.2:3b                    # For local testing
+```
+
+### Prompt Templates
+
+The system uses three types of LLM calls, each with its own prompt template:
+
+1. **Query Generation** (`queryPrompt`): Generates search queries based on the research topic
```

Comment:
Could apply the same as above to here, but with h4 tags.

Reply:
Done